File: Top Interview 150/Medium/FindKPairsWithSmallestSums.py
import heapq

class Solution(object):
    def kSmallestPairs(self, nums1, nums2, k):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :type k: int
        :rtype: List[List[int]]
        """
        len1 = len(nums1)
        len2 = len(nums2)
        sol  = []
        heap = []
        i = 0
        while i < k and i < len1:
            heapq.heappush(heap, (nums1[i] + nums2[0], i, 0))
            i += 1
        
        while k > 0 and heap:
            k -= 1
            _, i, j = heapq.heappop(heap)
            sol.append((nums1[i], nums2[j]))
            
            j += 1
            if j < len2:
                heapq.heappush(heap, (nums1[i] + nums2[j], i, j))
        
        return sol


if __name__ == "__main__":
    sol = Solution()
    # nums1, nums2, k = [1,1,2],  [1,2,3], 4
    # nums1, nums2, k = [1,7,11], [2,4,6], 3
    nums1, nums2, k = [1,2,4], [-1,1,2], 100
    print(sol.kSmallestPairs(nums1, nums2, k))

File: Top Interview 150/Medium/FindKPairsWithSmallestSums.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <set>

class Solution {
public:
    std::vector<std::vector<int>> kSmallestPairs(std::vector<int>& nums1, std::vector<int>& nums2, int k) {
        int len1 = nums1.size();
        int len2 = nums2.size();
        std::vector<std::vector<int>> sol;
        std::priority_queue<std::pair<int, std::pair<int, int>>, std::vector<std::pair<int, std::pair<int, int>>>, std::greater<std::pair<int, std::pair<int, int>>>> heap;
        
        for (int i = 0; i < k && i < len1; i++) {
            heap.push(std::make_pair(nums1[i] + nums2[0], std::make_pair(i, 0)));
        }
        
        while (k-- > 0 && !heap.empty()) {
            auto topElement = heap.top();
            int i = topElement.second.first;
            int j = topElement.second.second;
            heap.pop();
            sol.push_back({nums1[i], nums2[j]});
            
            if (++j < len2) {
                heap.push(std::make_pair(nums1[i] + nums2[j], std::make_pair(i, j)));
            }
        }
        
        return sol;
    }
};


int main () {
    Solution sol;
    std::vector<int> nums1{1,2,4};
    std::vector<int> nums2{-1,1,2};
    int k = 10;
    std::vector<std::vector<int>> result = sol.kSmallestPairs(nums1, nums2, k);
    
    std::cout << "[ ";
    for (const auto& r : result) {
        int i, j;
        i = r[0];
        j = r[1];
        std::cout << "(" << i << ", " << j << ") ";
    }
    std::cout << "]" << std::endl;
}